### DISCUSSION TITLE
Failing to establish a connection to server using amqp_client running on OTP26

### DISCUSSION THREAD

**Discussion by gmile** [rabbitmq/rabbitmq-server#8235] [Questions] [answered]
Hi, all!

We're giving OTP26 a try, and it looks due to changes related to SSL defaults in OTP26, the client code is no longer able to establish a connection with the server the same way it did with OTP25.

I've posted a full repro examples on [Elixirforum](https://elixirforum.com/t/issue-with-using-amqp-package-after-upgrading-to-otp26/55931) earlier today, but figured I'd open a discussion here too since the library we're using is essentially a wrapper around `amqp_client`, a source code for which is in current repo.

> **Note**: for reference, `amqp` uses `amqp_client` to make the connection in the following places in code:
> 
> * establishing what parameters will be used to make a connection: [link](https://github.com/pma/amqp/blob/0b630ac81be91b223e6152495e64e9a96a327b62/lib/amqp/connection.ex#L200-L219)
> * making a connection: [link](https://github.com/pma/amqp/blob/0b630ac81be91b223e6152495e64e9a96a327b62/lib/amqp/connection.ex#L184-L189)

In essence, I am failing to figure out correct options to pass to `amqp` (and subsequently to `amqp_client`) to bring the connection behaviour back to how it was with OTP25. Here's what I start with:

```elixir
# OTP25
AMQP.Connection.open("amqp://guest:guest@amqp-issue-rabbitmq")
{:ok, %AMQP.Connection{pid: #PID<0.241.0>}}
```

```elixir
# OTP26
AMQP.Connection.open("amqp://guest:guest@amqp-issue-rabbitmq")
# => {:error, {:auth_failure, 'Disconnected'}}
```

My hypothesis was that if I somehow find a way to disable SSL code path entirely, connection would work again. But I am failing miserably with a `:econnrefused` error:

```elixir
AMQP.Connection.open("amqp://guest:guest@amqp-issue-rabbitmq", [ssl_options: :none])
# => {:error, :econnrefused}
```

Any suggestions for what would be the correct way to bring the connection behaviour back to how it was with OTP25? 

P.S. I am happy to re-write the above example using pure `amqp_client` if that makes more sense to you - for now I went with what I could type faster.

**Comment by michaelklishin** [comment]
As mentioned in the [Erlang compatibility doc guide](https://www.rabbitmq.com/which-erlang.html), no currently released RabbitMQ version is compatible with Erlang 26.

**Comment by gmile** [comment]
@michaelklishin fair enough

Would you happen to know if there are any plans to have OTP26 support in the next minor version, for instance, in 3.12.1? Is there any on-going effort to implement OTP26 compatibility, e.g. a PR to keep track of?

**Reply by mkuratczyk** [reply]
https://github.com/rabbitmq/rabbitmq-server/issues/7845

At some point, we'll announce that 3.12.x is OTP26-compatible, but we don't know which exact version yet. The only known issues currently are dialyzer related, all tests pass on OTP26. The second problem is that we can't run the tests on Windows due to https://github.com/erlef/setup-beam/issues/189.

If you can test on OTP26, that'd be useful.

Any particular reason why you are interested in OTP26?

**Reply by gmile** [reply]
@mkuratczyk thanks, appreciate the pointers :+1: 

> Any particular reason why you are interested in OTP26?

No particular reason, no.

Some of our developers are likely to upgrade to OTP26 on their work stations soon once OTP26 becomes available in thier package managers. As an example, once this gets merged, https://github.com/Homebrew/homebrew-core/pull/131104, I assume some devs in the company will receive an automatic upgrade on the next `brew upgrade`.

After upgrading to OTP26, devs won't be able to start the application which needs a connection to RabbitMQ server.

However, upgrading is to OTP 26 is not anything crucial for us, we can definitely pin OTP 25 everywhere and wait for support in RabbitMQ, so no pressure from us :+1:

**Reply by mkuratczyk** [reply, marked as answer]
If they also update rabbitmq to the latest, it should work. We just don't feel comfortable saying it's supported yet, because we just recently made it pass the tests and as I mentioned, we can't even test on Windows yet. I am using OTP26 on my machine FWIW

**Reply by michaelklishin** [reply]
I use [`3.12.0-rc.3`](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.12.0-rc.3) on Erlang 26 as my daily driver. `3.12.0` is just around the corner, so please give this (possibly final) RC a shot :)

**Reply by gmile** [reply]
@michaelklishin thanks, I tried but ultimately failed to use `3.12.0-rc.3` 🤔 

1. Added this to under the `desp` function in `mix.exs`:

    ```elixir
    {:amqp_client,
     github: "rabbitmq/rabbitmq-server",
     tag: "v3.12.0-rc.3",
     subdir: "deps/amqp_client",
     override: true},
    ```

2. Attempted to start the app (note I am still using OTP 25, but this should be irrelevant), which failed:

    ```
    $ iex -S mix
    Erlang/OTP 25 [erts-13.2.2] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit:ns] [dtrace]

     GEN    clean-profile
     GEN    clean-app
     GEN    coverdata-clean
     GEN    clean-dist
     GEN    clean-ct-logs-archive
     GEN    clean-rabbitmq-deps.mk
     DEP    credentials_obfuscation ()
    Error: Unknown or invalid dependency: credentials_obfuscation.
    make: *** [erlang.mk:5105: /Users/eugene/work/contractbook/api/deps/amqp_client/deps/credentials_obfuscation] Error 78
    ** (Mix) Could not compile dependency :amqp_client, "make clean && make" command failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile amqp_client", update it with "mix deps.update amqp_client" or clean it with "mix deps.clean amqp_client"
    $
    ```

Any chance to have RCs of `amqp_client` be published on Hex to make the installation easier? Otherwise, I think I'll wait until the final release of 3.12 to try it out with OTP26.

**Reply by michaelklishin** [reply]
Previews are also published to Hex.pm. Apparently the publishing job has been failing since 3.11.14.

**Reply by gmile** [reply]
@michaelklishin I've looked briefly at the workflows in the `rabbitmq/rabbitmq-server`, as well at other repos - failed to see anything related to releasing of the client code. I'd be happy to see if I can do anything to fix releasing of the client code - any clues where I can find it?

**Reply by lukebakken** [reply]
@gmile I have fixed it here - https://github.com/rabbitmq/rabbitmq-server/pull/8266

Publishing to hex.pm happens via our internal CI systems. The next release should work correctly.

**Reply by gmile** [reply]
While recent RC for `amqp_client` are not yet published, I tried adding a dependency using a 349e0a5040b167bc93fc02c8d91bc5551d4d1379 ref directly, but failed.

Having the following as an Elixir script:

```elixir
# test.exs

Mix.install([
  :amqp,
  {:amqp_client,
   github: "rabbitmq/rabbitmq-server",
   ref: "9884238b31ac8bcd2b771a045a200200373445ec",
   subdir: "deps/amqp_client",
   override: true}
])

IO.puts("It worked")
```

Running `elixir test.exs` results in a compilation error related to `rabbit_common` availability:

```
== Compilation error in file lib/amqp/core.ex ==
** (ArgumentError) lib file rabbit_common/include/rabbit_framing.hrl could not be found
    (elixir 1.14.4) lib/record/extractor.ex:41: Record.Extractor.from_lib_file/1
    (elixir 1.14.4) lib/record/extractor.ex:18: Record.Extractor.from_or_from_lib_file/1
    (elixir 1.14.4) lib/record/extractor.ex:5: Record.Extractor.extract/2
    lib/amqp/core.ex:9: (module)
```

<details>
<summary>Full output</summary>

```
MIX_BUILD_ROOT=/tmp/amqp-issue-builds elixir amqp.exs
* Getting amqp_client (https://github.com/rabbitmq/rabbitmq-server.git - 9884238b31ac8bcd2b771a045a200200373445ec)
remote: Enumerating objects: 371287, done.
remote: Counting objects: 100% (1931/1931), done.
remote: Compressing objects: 100% (903/903), done.
remote: Total 371287 (delta 1166), reused 1650 (delta 1022), pack-reused 369356
Resolving Hex dependencies...
Resolution completed in 0.022s
New:
  amqp 2.1.2
* Getting amqp (Hex package)
erlang.mk:26: Please upgrade to GNU Make 4 or later: https://erlang.mk/guide/installation.html
 GEN    clean-profile
 GEN    clean-app
 GEN    coverdata-clean
 GEN    clean-dist
 GEN    clean-ct-logs-archive
 GEN    clean-rabbitmq-deps.mk
erlang.mk:26: Please upgrade to GNU Make 4 or later: https://erlang.mk/guide/installation.html
 DEP    credentials_obfuscation (3.4.0)
No beam files found.
Recompile: src/rebar
Recompile: src/rebar_abnfc_compiler
Recompile: src/rebar_app_utils
Recompile: src/rebar_appups
Recompile: src/rebar_asn1_compiler
Recompile: src/rebar_base_compiler
Recompile: src/rebar_cleaner
Recompile: src/rebar_config
Recompile: src/rebar_core
Recompile: src/rebar_cover_utils
Recompile: src/rebar_ct
src/rebar_ct.erl:291:30: Warning: crypto:rand_uniform/2 is deprecated; use rand:uniform/1 instead
%  291|     Random = integer_to_list(crypto:rand_uniform(0, 10000)),
%     |                              ^

Recompile: src/rebar_deps
Recompile: src/rebar_dia_compiler
Recompile: src/rebar_dialyzer
Recompile: src/rebar_edoc
Recompile: src/rebar_erlc_compiler
Recompile: src/rebar_erlydtl_compiler
Recompile: src/rebar_escripter
Recompile: src/rebar_eunit
src/rebar_eunit.erl:282:20: Warning: crypto:rand_uniform/2 is deprecated; use rand:uniform/1 instead
%  282|             Seed = crypto:rand_uniform(1, 65535),
%     |                    ^

Recompile: src/rebar_file_utils
Recompile: src/rebar_getopt
Recompile: src/rebar_lfe_compiler
Recompile: src/rebar_log
Recompile: src/rebar_metacmds
Recompile: src/rebar_mustache
Recompile: src/rebar_neotoma_compiler
Recompile: src/rebar_otp_app
Recompile: src/rebar_otp_appup
Recompile: src/rebar_port_compiler
Recompile: src/rebar_proto_compiler
Recompile: src/rebar_proto_gpb_compiler
Recompile: src/rebar_protobuffs_compiler
Recompile: src/rebar_qc
Recompile: src/rebar_rand_compat
Recompile: src/rebar_rel_utils
Recompile: src/rebar_reltool
Recompile: src/rebar_require_vsn
Recompile: src/rebar_shell
Recompile: src/rebar_subdirs
Recompile: src/rebar_templater
Recompile: src/rebar_upgrade
Recompile: src/rebar_utils
Recompile: src/rebar_xref
Recompile: src/rmemo
==> rebar (compile)
==> rebar (escriptize)
Congratulations! You now have a self-contained script called "rebar" in
your current working directory. Place this script anywhere in your path
and you can use rebar to build OTP-compliant apps.
/Users/eugene/Library/Caches/mix/installs/elixir-1.14.4-erts-13.2.2/e642fa749c90c0622598ef4cd6315d03/deps/amqp_client/deps/amqp_client
 DEP    thoas (1.0.0)
 DEP    recon (2.5.3)
 DEPEND thoas.d
 ERLC   thoas.erl thoas_decode.erl thoas_encode.erl
Old inliner: threshold=0 functions=[{array,6},
                                    {object,6},
                                    {token_error,3},
                                    {token_error,2},
                                    {throw_error,2},
                                    {continue,6}]
Old inliner: threshold=0 functions=[{float,1},
                                    {integer,1},
                                    {error_invalid_byte_error,2}]
 APP    thoas.app.src
 DEPEND recon.d
 ERLC   recon.erl recon_alloc.erl recon_lib.erl recon_map.erl recon_rec.erl recon_trace.erl
Old inliner: threshold=0 functions=[{insert,2},{merge,2}]
 APP    recon.app.src
 DEPEND credentials_obfuscation.d
 ERLC   credentials_obfuscation.erl credentials_obfuscation_app.erl credentials_obfuscation_pbe.erl credentials_obfuscation_sup.erl credentials_obfuscation_svc.erl
 APP    credentials_obfuscation.app.src
 GEN    include/rabbit_framing.hrl
 GEN    src/rabbit_framing_amqp_0_8.erl
 GEN    src/rabbit_framing_amqp_0_9_1.erl
 DEPEND rabbit_common.d
 ERLC   app_utils.erl code_version.erl credit_flow.erl delegate.erl delegate_sup.erl file_handle_cache.erl file_handle_cache_stats.erl gen_server2.erl mirrored_supervisor_locks.erl mnesia_sync.erl pmon.erl priority_queue.erl rabbit_amqp_connection.erl rabbit_amqqueue_common.erl rabbit_auth_backend_dummy.erl rabbit_auth_mechanism.erl rabbit_authn_backend.erl rabbit_authz_backend.erl rabbit_basic_common.erl rabbit_binary_generator.erl rabbit_binary_parser.erl rabbit_cert_info.erl rabbit_channel_common.erl rabbit_command_assembler.erl rabbit_control_misc.erl rabbit_core_metrics.erl rabbit_data_coercion.erl rabbit_date_time.erl rabbit_env.erl rabbit_error_logger_handler.erl rabbit_event.erl rabbit_framing.erl rabbit_framing_amqp_0_8.erl rabbit_framing_amqp_0_9_1.erl rabbit_heartbeat.erl rabbit_http_util.erl rabbit_json.erl rabbit_log.erl rabbit_misc.erl rabbit_msg_store_index.erl rabbit_net.erl rabbit_nodes_common.erl rabbit_numerical.erl rabbit_password.erl rabbit_password_hashing.erl rabbit_password_hashing_md5.erl rabbit_password_hashing_sha256.erl rabbit_password_hashing_sha512.erl rabbit_pbe.erl rabbit_peer_discovery_backend.erl rabbit_policy_validator.erl rabbit_queue_collector.erl rabbit_registry.erl rabbit_registry_class.erl rabbit_resource_monitor_misc.erl rabbit_runtime.erl rabbit_runtime_parameter.erl rabbit_semver.erl rabbit_semver_parser.erl rabbit_ssl_options.erl rabbit_types.erl rabbit_writer.erl supervisor2.erl vm_memory_monitor.erl worker_pool.erl worker_pool_sup.erl worker_pool_worker.erl
 APP    rabbit_common
 DEPEND amqp_client.d
 ERLC   amqp_auth_mechanisms.erl amqp_channel.erl amqp_channel_sup.erl amqp_channel_sup_sup.erl amqp_channels_manager.erl amqp_client.erl amqp_connection.erl amqp_connection_sup.erl amqp_connection_type_sup.erl amqp_direct_connection.erl amqp_direct_consumer.erl amqp_gen_connection.erl amqp_gen_consumer.erl amqp_main_reader.erl amqp_network_connection.erl amqp_rpc_client.erl amqp_rpc_server.erl amqp_selective_consumer.erl amqp_ssl.erl amqp_sup.erl amqp_uri.erl amqp_util.erl rabbit_routing_util.erl uri_parser.erl
 APP    amqp_client
==> amqp
Compiling 16 files (.ex)

== Compilation error in file lib/amqp/core.ex ==
** (ArgumentError) lib file rabbit_common/include/rabbit_framing.hrl could not be found
    (elixir 1.14.4) lib/record/extractor.ex:41: Record.Extractor.from_lib_file/1
    (elixir 1.14.4) lib/record/extractor.ex:18: Record.Extractor.from_or_from_lib_file/1
    (elixir 1.14.4) lib/record/extractor.ex:5: Record.Extractor.extract/2
    lib/amqp/core.ex:9: (module)
could not compile dependency :amqp, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile amqp", update it with "mix deps.update amqp" or clean it with "mix deps.clean amqp"
```
</details>

The above is probably expected, as it looks like while `amqp_client` compiles just fine (and also compiles `rabbit_common` too) - `rabbit_common` is not really available to `amqp` (which [needs](https://github.com/pma/amqp/blob/0b630ac81be91b223e6152495e64e9a96a327b62/lib/amqp/core.ex#L105) it).

Unfortunately, adding `rabbit_common` explicitly doesn't seem to help work around the issue either, though sadly which I think makes sense as `rabbit_common` dependency is not really resolved in this case:

```elixir
# test.exs

Mix.install([
  :amqp,
  {:amqp_client,
   github: "rabbitmq/rabbitmq-server",
   ref: "9884238b31ac8bcd2b771a045a200200373445ec",
   subdir: "deps/amqp_client",
   override: true},
  {:rabbit_common,
   github: "rabbitmq/rabbitmq-server",
   ref: "9884238b31ac8bcd2b771a045a200200373445ec",
   subdir: "deps/rabbit_common",
   override: true}
])

IO.puts("It worked")
```

**Reply by michaelklishin** [reply]
`3.11.17` and `3.12.0-rc.4` releases of `amqp_client` should now be available from Hex.pm. We will not advertise this for 3.11 but it is almost as compatible with Erlang 26 as 3.12.0 is/will be.
And in the specific case of `amqp_client`, they should be equally compatible.

Please give both of them a try :)

**Reply by gmile** [reply]
@michaelklishin thanks, I did try amqp_client `3.12.0-rc.4` but forgot to report back - under OTP 26, it now is able to connect to the RabbitMQ server, thanks! I will see if downgrading to `3.11.17` works too, and will report back
